File: rubber.fs
\ rubber.fs -- rubber.scm --> rubber.fs

\ Translator: Michael Scholz <mi-scholz@users.sourceforge.net>
\ Created: 06/01/06 05:32:57
\ Changed: 14/04/28 03:52:17
\
\ @(#)rubber.fs	1.15 4/28/14

\ original comments are taken from rubber.scm [ms]
\
\ ;;; rubber.scm: rubber-sound stretches or contracts a sound (in time)
\ ;;;   (rubber-sound 1.5) makes it 50% longer
\ ;;;   rubber-sound looks for stable portions and either inserts or deletes
\ ;;;     periods period length is determined via autocorrelation

require clm
require marks

8             value zeros-checked
10.0          value extension
*fth-verbose* value show-details

hide
\ ;;; remove anything below 16Hz
\ ;;; extend (src by 1/extension)
\ ;;; collect upward zero-crossings
\ ;;;   collect weights for each across next zeros-checked crossings
\ ;;;   sort by least weight
\ ;;;   ramp (out or in) and check if done

: derumble-sound { snd chn -- }
	\ ;; remove rumbles and DC etc (since we're using zero crossings
	\ ;;   to find period starts)
	snd chn #f framples { old-len }
	snd srate { sr }
	#( 0.0 0.0
	   16.0 f2* sr f/ 0.0
	   20.0 f2* sr f/ 1.0
	   1.0 1.0 ) ( flt-lst )
	    2.0 old-len sr min flog 2.0 flog f/ fceil ( pow2 )
	    f** f>s ( fftlen ) snd chn #f undef filter-sound drop
	old-len snd chn set-framples drop
;

: sample-sound { snd chn -- }
	\ ;; prepare sound for analysis by interpolating samples
	extension 1.0 f<> if
		extension 1/f 1.0 snd chn #f src-sound drop
	then
;

: unsample-sound { snd chn -- }
	\ ;; undo earlier interpolation
	extension 1.0 f<> if
		extension 1.0 snd chn #f src-sound drop
	then
;

: crossings { snd chn -- n }
	\ ;; return number of upward zero crossings that don't look like silence
	0 0 { crosses last-cross }
	0 snd chn 1 #f make-sampler { sr0 }
	sr0 next-sample { samp0 }
	0.0 { sum }
	extension 0.001 f* { silence }
	snd chn #f framples 0 ?do
		sr0 next-sample { samp1 }
		samp0 f0<=
		samp1 f0>          &&
		i last-cross - 4 > &&
		sum silence f>     && if
			crosses 1+ to crosses
			i to last-cross
			0.0 to sum
		then
		samp0 fabs sum f+ to sum
		samp1 to samp0
	loop
	crosses
;

: env-add { s0 s1 samps snd chn -- vct }
	1.0 { x }
	samps 1/f { xinc }
	s0 snd chn 1 #f make-sampler { sr0 }
	s1 snd chn 1 #f make-sampler { sr1 }
	samps 0.0 make-vct map!
		sr0 next-sample x f*
		    sr1 next-sample  1.0 x f-  f*  f+ ( val )
		x xinc f+ to x
		( val )
	end-map ( data )
;

: rubber-cb { stretch snd chn -- prc; self -- }
	0 proc-create ( prc ) chn , snd , stretch ,
 does> { self -- }
	self @ { chn }
	self cell+ @ { snd }
	self 2 cells + @ { stretch }
	\ ;; prepare sound (get rid of low freqs, resample)
	snd chn derumble-sound
	snd chn sample-sound
	snd chn crossings { crosses }
	crosses :initial-element 0 make-array { cross-samples }
	crosses :initial-element 0 make-array { cross-weights }
	crosses :initial-element 0 make-array { cross-marks }
	crosses :initial-element 0 make-array { cross-periods }
	\ ;; get cross points (sample numbers)
	0 snd chn 1 #f make-sampler { sr0 }
	sr0 next-sample { samp0 }
	0.0 { sum }
	0 { last-cross }
	extension 0.001 f* { silence }
	snd chn #f framples 0 ?do
		sr0 next-sample { samp1 }
		samp0 f0<=
		samp1 f0>           &&
		i last-cross - 40 > &&
		sum silence f>      && if
			i to last-cross
			0.0 to sum
			cross-samples i cycle-set!
		then
		samp0 fabs sum f+ to sum
		samp1 to samp0
	loop
	\ ;; now run through crosses getting period match info
	crosses 1- 0 ?do
		cross-samples i array-ref { start }
		0 { autolen }
		2.0 extension snd srate 40.0 f/ f*
		flog 2.0 flog f/ fceil ( pow2 ) f** f>s { fftlen }
		fftlen 4 / { len4 }
		start snd chn 1 #f make-sampler { rd }
		fftlen 0.0 make-vct map! rd next-sample end-map { data }
		data autocorrelate drop
		len4 1 ?do
			data i    vct-ref data i 1+  vct-ref f<
			data i 1+ vct-ref data i 2 + vct-ref f> && if
				i 2* to autolen
				leave
			then
		loop
		start autolen + { next-start }
		i 1+ { min-i }
		cross-samples min-i array-ref next-start - abs { min-samps }
		crosses i zeros-checked + min i 2 + ?do
			cross-samples i array-ref next-start - abs { dist }
			dist min-samps < if
				dist to min-samps
				i to min-i
			then
		loop
		min-i { current-mark }
		0.0 { current-min }
		cross-samples current-mark array-ref { s1 }
		start snd chn 1 #f make-sampler { sr0 }
		s1 snd chn 1 #f make-sampler { sr1 }
		0.0 0.0 { ampsum diffsum }
		autolen 0 ?do
			sr0 next-sample { samp0 }
			sr1 next-sample { samp1 }
			samp0 fabs ampsum f+ to ampsum
			samp1 samp0 f- fabs diffsum f+ to diffsum
		loop
		diffsum f0= if
			0.0
		else
			diffsum ampsum f/
		then to current-min
		current-min f2/ fround->s to min-samps
		current-mark zeros-checked i + min crosses 1- min { top }
		top i 1+ ?do
			0.0 { wgt }
			cross-samples i array-ref { s1 }
			start snd chn 1 #f make-sampler { sr0 }
			s1 snd chn 1 #f make-sampler { sr1 }
			0.0 0.0 { ampsum diffsum }
			autolen 0 ?do
				sr0 next-sample { samp0 }
				sr1 next-sample { samp1 }
				samp0 fabs ampsum f+ to ampsum
				samp1 samp0 f- fabs diffsum f+ to diffsum
			loop
			diffsum f0= if
				0.0
			else
				diffsum ampsum f/
			then to wgt
			wgt min-samps f< if
				wgt f>s to min-samps
				i to min-i
			then
		loop
		current-mark min-i <> if
			\ ;; these are confused, so effectively erase them
			cross-weights i 1000.0 array-set!
		else
			cross-weights i current-min array-set!
			cross-marks i current-mark array-set!
			cross-periods i cross-samples current-mark array-ref
			cross-samples i array-ref -  array-set!
		then
	loop
	\ ;; now sort weights to scatter the changes as evenly as possible
	snd chn #f framples { len }
	stretch 1.0 f> { adding }
	stretch 1.0 f- fabs len f* floor f>s { samps }
	adding if
		samps
	else
		len samps 2* min
	then { needed-samps }
	0 { handled }
	#() { edits }
	begin
		\ ;; need to find (more than) enough splice pts to delete samps
		-1      { best-mark }
		handled { old-handled }
		0       { cur }
		cross-weights 0 array-ref { curmin }
		cross-weights each { val }
			val curmin f< if
				i to cur
				val to curmin
			then
		end-each
		cur to best-mark
		cross-periods best-mark array-ref handled + to handled
		handled needed-samps <
		handled needed-samps - needed-samps old-handled - < || if
			edits best-mark array-push drop
		then
		cross-weights best-mark 1000.0 array-set!
		edits length crosses =
		handled needed-samps >= ||
	until
	edits length crosses >= if
		needed-samps handled f/ fceil f>s
	else
		1
	then { mult }
	0 { changed-len }
	edits each { best-mark }
		changed-len samps > ?leave
		cross-samples best-mark array-ref { beg }
		cross-samples 
		    cross-marks best-mark array-ref
		    array-ref { next-beg }
		cross-periods best-mark array-ref { len }
		len 0> if
			adding if
				beg next-beg len snd chn env-add { new-samps }
				show-details if
					beg "%d:%d" #( i len extension f/ f>s )
					    string-format
					    snd chn add-named-mark drop
				then
				beg len new-samps snd chn #f insert-samples drop
				mult 1 ?do
					beg len i * + len new-samps
					    snd chn #f insert-samples drop
				loop
				len mult * changed-len + to changed-len
				cross-samples each { curbeg }
					curbeg beg > if
						cross-samples i
						    curbeg len + array-set!
					then
				end-each
			else
				beg snd chn #f framples >= if
					"trouble at %d: %d of %d"
					    #( i beg snd chn #f framples )
					    clm-message
				then
				show-details if
					beg 1- "%d:%d"
					    #( i len extension f/ f>s )
					    string-format
					    snd chn add-named-mark drop
				then
				beg len snd chn #f delete-samples drop
				changed-len len + to changed-len
				beg len + { end }
				cross-samples each { curbeg }
					curbeg beg > if
						curbeg end < if
							cross-periods i 0
							    array-set!
						else
							cross-samples i
							    curbeg len -
							    array-set!
						then
					then
				end-each
			then
		then
	end-each
	show-details if
		"wanted: %d, got %d" #( samps changed-len ) clm-message
	then
	\ ;; and return to original srate
	snd chn unsample-sound
	show-details if
		snd chn 0 framples { frms0 }
		snd chn undef framples { frms }
		"%d -> %d (%d)"
		    #( frms0  frms  frms0 stretch f* floor f>s ) clm-message
	then
;
set-current

: rubber-sound { stretch snd chn -- res }
	"%s %s" #( stretch get-func-name ) string-format { origin }
	stretch snd chn rubber-cb origin as-one-edit
;
previous

\ rubber.fs ends here

File: rubber.rb
# rubber.rb -- Translation of rubber.scm

# Translator/Author: Michael Scholz <mi-scholz@users.sourceforge.net>
# Created: 03/02/28 03:04:03
# Changed: 14/11/23 06:14:52

# module Rubber (see rubber.scm)
#  add_named_mark(samp, name, snd, chn)
#  derumble_sound(snd, chn)
#  sample_sound(snd, chn)
#  unsample_sound(snd, chn)
#  crossings()
#  env_add(s0, s1, samps)
#  rubber_sound(stretch, snd, chn)

require "clm"

module Rubber
  doc "#{self.class} #{self.name} is the translation of rubber.scm.\n"
  
  $zeros_checked = 8
  $extension = 10.0
  $show_details = false
  
  def add_named_mark(samp, name, snd = false, chn = false)
    m = add_mark(samp.round, snd, chn)
    set_mark_name(m, name)
    m
  end

  def derumble_sound(snd = false, chn = false)
    old_length = framples(snd, chn)
    pow2 = (log([old_length, srate(snd)].min) / log(2)).ceil
    fftlen = (2 ** pow2).round
    flt_env = [0.0, 0.0, 32.0 / srate(snd), 0.0,
               40.0 / srate(snd), 1.0, 1.0, 1.0]
    filter_sound(flt_env, fftlen, snd, chn)
    set_framples(old_length, snd, chn)
  end

  def sample_sound(snd = false, chn = false)
    if $extension != 1.0
      src_sound(1.0 / $extension, 1.0, snd, chn)
    end
  end

  def unsample_sound(snd = false, chn = false)
    if $extension != 1.0
      src_sound($extension, 1.0, snd, chn)
    end
  end

  def crossings
    crosses = 0
    sr0 = make_sampler(0)
    samp0 = next_sample(sr0)
    len = framples()
    sum = 0.0
    last_cross = 0
    silence = $extension * 0.001
    (0...len).each do |i|
      samp1 = next_sample(sr0)
      if samp0 <= 0.0 and
          samp1 > 0.0 and
          (i - last_cross) > 4 and
          sum > silence
        crosses += 1
        last_cross = i
        sum = 0.0
      end
      sum += samp0.abs
      samp0 = samp1
    end
    crosses
  end

  def env_add(s0, s1, samps)
    data = make_vct(samps.round)
    x = 1.0
    xinc = 1.0 / samps
    sr0 = make_sampler(s0.round)
    sr1 = make_sampler(s1.round)
    (0...samps).each do |i|
      data[i] = x * next_sample(sr0) + (1.0 - x) * next_sample(sr1)
      x += xinc
    end
    data
  end

  def rubber_sound(stretch, snd = false, chn = false)
    as_one_edit(lambda do | |
                  derumble_sound(snd, chn)
                  sample_sound(snd, chn)
                  crosses = crossings()
                  cross_samples = make_vct(crosses)
                  cross_weights = make_vct(crosses)
                  cross_marks = make_vct(crosses)
                  cross_periods = make_vct(crosses)
                  sr0 = make_sampler(0, snd, chn)
                  samp0 = next_sample(sr0)
                  len = framples()
                  sum = 0.0
                  last_cross = 0
                  cross = 0
                  silence = $extension * 0.001
                  (0...len).each do |i|
                    samp1 = next_sample(sr0)
                    if samp0 <= 0.0 and
                        samp1 > 0.0 and
                        (i - last_cross) > 4 and
                        sum > silence
                      last_cross = i
                      sum = 0.0
                      cross_samples[cross] = i
                      cross += 1
                    end
                    sum += samp0.abs
                    samp0 = samp1
                  end
                  (0...(crosses - 1)).each do |i|
                    start = cross_samples[i]
                    autolen = 0
                    s0 = start
                    pow2 = (log($extension * (srate() / 40.0)) / log(2)).ceil
                    fftlen = (2 ** pow2).round
                    len4 = fftlen / 4
                    data = make_vct(fftlen)
                    reader = make_sampler(s0.round)
                    (0...fftlen).each do |j|
                      data[j] = next_sample(reader)
                    end
                    autocorrelate(data)
                    autolen = 0
                    (1...len4).detect do |j|
                      if data[j] < data[j + 1] and data[j + 1] > data[j + 2]
                        autolen = j * 2
                        true
                      end
                    end
                    next_start = start + autolen
                    min_i = i + 1
                    min_samps = (cross_samples[min_i] - next_start).abs
                    ((i + 2)...[crosses, i + $zeros_checked].min).each do |k|
                      dist = (cross_samples[k] - next_start).abs
                      if dist < min_samps
                        min_samps = dist
                        min_i = k
                      end
                    end
                    current_mark = min_i
                    current_min = 0.0
                    s0 = start
                    s1 = cross_samples[current_mark]
                    len = autolen
                    sr0 = make_sampler(s0.round)
                    sr1 = make_sampler(s1.round)
                    ampsum = 0.0
                    diffsum = 0.0
                    (0...len).each do |dummy|
                      samp0 = next_sample(sr0)
                      samp1 = next_sample(sr1)
                      ampsum += samp0.abs
                      diffsum += (samp1 - samp0).abs
                    end
                    if diffsum == 0.0
                      current_min = 0.0
                    else
                      current_min = diffsum / ampsum
                    end
                    min_samps = 0.5 * current_min
                    top = [crosses - 1, current_mark, i + $zeros_checked].min
                    ((i + 1)...top).each do |k|
                      wgt = 0.0
                      s0 = start
                      s1 = cross_samples[k]
                      len = autolen
                      sr0 = make_sampler(s0.round)
                      sr1 = make_sampler(s1.round)
                      ampsum = 0.0
                      diffsum = 0.0
                      (0...len).each do |dummy|
                        samp0 = next_sample(sr0)
                        samp1 = next_sample(sr1)
                        ampsum += samp0.abs
                        diffsum += (samp1 - samp0).abs
                      end
                      if diffsum == 0.0
                        wgt = 0.0
                      else
                        wgt = diffsum / ampsum
                      end
                      if wgt < min_samps
                        min_samps = wgt
                        min_i = k
                      end
                    end
                    unless current_mark == min_i
                      cross_weights[i] = 1000.0
                    else
                      cross_weights[i] = current_min
                      cross_marks[i] = current_mark
                      cross_periods[i] = cross_samples[current_mark] -
                                         cross_samples[i]
                    end
                  end
                  len = framples(snd, chn)
                  adding = (stretch > 1.0)
                  samps = ((stretch - 1.0).abs * len).round
                  needed_samps = (adding ? samps : [len, samps * 2].min)
                  handled = 0
                  mult = 1
                  curs = 0
                  weigths = cross_weights.length
                  edits = make_vct(weigths)
                  until curs == weigths or handled >= needed_samps
                    best_mark = -1
                    old_handled = handled
                    cur = 0
                    curmin = cross_weights[0]
                    cross_weights.each_with_index do |cross_w, i|
                      if cross_w < curmin
                        cur = i
                        curmin = cross_w
                      end
                    end
                    best_mark = cur
                    handled += cross_periods[best_mark].round
                    if (handled < needed_samps) or
                       ((handled - needed_samps) < (needed_samps - old_handled))
                      edits[curs] = best_mark
                      curs += 1
                    end
                    cross_weights[best_mark] = 1000.0
                  end
                  mult = (needed_samps / handled).ceil if curs >= weigths
                  changed_len = 0
                  weights = cross_weights.length
                  (0...curs).detect do |i|
                    best_mark = edits[i].round
                    beg = cross_samples[best_mark].to_i
                    next_beg = cross_samples[cross_marks[best_mark].round]
                    len = cross_periods[best_mark].to_i
                    if len > 0
                      if adding
                        new_samps = env_add(beg, next_beg, len)
                        if $show_details
                          add_named_mark(beg,
                                         format("%d:%d",
                                                i, (len / $extension).round))
                        end
                        insert_samples(beg, len, new_samps)
                        if mult > 1
                          (1...mult).each do |k|
                            insert_samples(beg + k * len, len, new_samps)
                          end
                        end
                        changed_len = changed_len + mult * len
                        (0...weights).each do |j|
                          curbeg = cross_samples[j]
                          if curbeg > beg
                            cross_samples[j] = curbeg + len
                          end
                        end
                      else
                        frms = framples()
                        if beg >= frms
                          Snd.display("trouble at %d: %d of %d", i, beg, frms)
                        end
                        if $show_details
                          add_named_mark(beg - 1,
                                         format("%d:%d",
                                                i, (len / $extension).round))
                        end
                        delete_samples(beg, len)
                        changed_len += len
                        fin = beg + len
                        (0...weights).each do |j|
                          curbeg = cross_samples[j]
                          if curbeg > beg
                            if curbeg < fin
                              cross_periods[j] = 0
                            else
                              cross_samples[j] = curbeg - len
                            end
                          end
                        end
                      end
                    end
                    changed_len > samps
                  end
                  if $show_details
                    Snd.display("wanted: %d, got %d",
                                samps.round, changed_len.round)
                  end
                  unsample_sound(snd, chn)
                  if $show_details
                    frms0 = framples(snd, chn, 0)
                    Snd.display("%f -> %f (%f)",
                                frms0,
                                framples(snd,chn),
                                (stretch * frms0).round)
                  end
                end,
                format("rubber_sound(%f, %p, %p,", stretch, snd, chn))
  end
end

include Rubber

# rubber.rb ends here
